lib/bosh/BoshDirectorClient.js

```diff
@@ -16,27 +16,133 @@ const NotFound = errors.NotFound;
 const BadRequest = errors.BadRequest;
 
 class BoshDirectorClient extends HttpClient {
-  constructor(boshDirectorConfig) {
-    const boshDirector = boshDirectorConfig || config.director;
+  constructor() {
+    // activeConfigs - directors supporting lifecycle oprations
+    let activeConfigs = BoshDirectorClient.getActiveConfigs();
+    // directorConfig - directors supporting 'create' oprations
+    let coreConfigs = BoshDirectorClient.getCoreConfigs(activeConfigs);
+    // inactiveConfigs - directors for OutOfBand backup scenario
+    let inactiveConfigs = BoshDirectorClient.getInactiveConfigs();
+    let config = BoshDirectorClient.getPrimaryConfig(coreConfigs);
     super({
-      baseUrl: boshDirector.url,
-      auth: {
-        user: boshDirector.username,
-        pass: boshDirector.password
-      },
       headers: {
         Accept: 'application/json,text/plain;q=0.9'
       },
       followRedirect: false,
-      rejectUnauthorized: !boshDirector.skip_ssl_validation
+      rejectUnauthorized: !config.skip_ssl_validation
+    });
+    this.activeConfigs = activeConfigs;
+    this.coreConfigs = coreConfigs;
+    this.inactiveConfigs = inactiveConfigs;
+    this.config = config;
+    logger.info('Active Configs', this.activeConfigs);
+    logger.info('Core Configs', this.coreConfigs);
+    logger.info('Inactive Configs', this.inactiveConfigs);
+    logger.info('Create Config', this.config);
+    this.primary = this.config.primary || false;
+    this.uuid = this.config.uuid;
+    this.cpi = this.config.cpi;
+    this.infrastructure = this.config.infrastructure;
+    this.populateCache();
+  }
+
+  clearCache() {
+    this.cache = {};
+  }
+
+  static getActiveConfigs() {
+    return _.filter(config.directors, function (director) {
+      return director.primary;
+    });
+  }
+
+  static getCoreConfigs(activeConfigs) {
+    return _.filter(activeConfigs, function (director) {
+      return director.supportCreate;
+    });
+  }
+
+  static getInactiveConfigs() {
+    return _.filter(config.directors, function (director) {
+      return !director.primary;
     });
-    this.uuid = boshDirector.uuid;
-    this.cpi = boshDirector.cpi;
+  }
+
+  static getPrimaryConfig(coreConfigs) {
+    return _.sample(coreConfigs);
+  }
+
+  getConfigByName(name) {
+    return _.head(_.filter(config.directors, (director) => director.name === name));
+  }
+
+  populateCache() {
+    this.clearCache();
+    return this
+      .getDeployments()
+      .tap(() => logger.debug('Cached Deployments:', this.cache));
+  }
+
+  addInactiveToCache() {
+    return this
+      .getDeployments(true)
+      .tap(() => logger.debug('Cached Deployments:', this.cache));
+  }
+
+  deleteCacheEntry(deploymentName) {
```

Comment:
method not required. No harm in retaining a mapping in cache entry post deployment deletion, this will anyways get cleared on restart and even if we did find mapping with a director, the bosh director throws deployment not found message anyways.